LCD_BED_TRAMMING fixes

Farva42 opened the pull request:
### Description
This allows LCD_BED_TRAMMING to work with a stowable probe like MAG_MOUNTED_PROBE, SERVO_PROBE, or TOUCH_MI_PROBE. Before if the probe was not deployed before initiating tramming the printer would not attempt to deploy the probe and loop continually raising Z till it hit Z max. 
The no_mem was added to reduce excessive movement before and after probe deploy/stow moves.

This also fixes bed tramming ignoring probe offsets which can cause issues with the printer trying to probe off the print bed.

### Requirements

LCD_BED_TRAMMING to be enabled  with BED_TRAMMING_USE_PROBE to see the offset issue.
both the above and a prob covered in HAS_STOWABLE_PROBE for that fix 

### Benefits

Lets printers with HAS_STOWABLE_PROBE use LCD_BED_TRAMMING easily and fixes the probe offset issue.


### Related Issues
(https://github.com/MarlinFirmware/Marlin/issues/24728)

Farva42 committed:
Fix LCD_BED_TRAMMING

Fix probe offset not being applied & enable non BLtouch probes to be used without needing to deploy the probe prior to initiating bed tramming

Farva42 committed:
Update menu_bed_corners.cpp

Farva42 committed:
Fix bad definition

Farva42 committed:
Fix un-garded bool use

Farva42 commented:
Bah sneaky BLtouch flagging HAS_STOWABLE_PROBE via being a Servo probe. I need to guard the probe.deploy and stows from getting called with a bltouch enabled before this is good.

Farva42 committed:
Fix BLTOUCH guarding

thinkyhead committed:
add CI test

thinkyhead commented on Marlin/src/lcd/menu/menu_bed_corners.cpp:
```diff
@@ -208,9 +208,15 @@ static void _lcd_level_bed_corners_get_next_position() {
-      , []{ queue.inject(TERN(HAS_LEVELING, F("G29N"), FPSTR(G28_STR))); ui.return_to_status(); }
-      , TERN(HAS_LEVELING, ui.goto_previous_screen_no_defer, []{})
+        TERN(HAS_LEVELING, GET_TEXT_F(MSG_BUTTON_DONE), nullptr)
+      , []{ TERN(HAS_LEVELING, []{queue.inject(F("G29N")); corner_probing_done = true;}, queue.inject(FPSTR(G28_STR))); ui.goto_previous_screen_no_defer();}
```
- The code wrapped in `[]{}` will not be executed. It just defines an anonymous lambda.
- This screen is the final one, and it should go back to Status so leveling / homing are reported.

thinkyhead commented on Marlin/src/lcd/menu/menu_bed_corners.cpp:
```diff
@@ -208,9 +208,15 @@ static void _lcd_level_bed_corners_get_next_position() {
+        TERN(HAS_LEVELING, GET_TEXT_F(MSG_BUTTON_DONE), nullptr)
+      , []{ TERN(HAS_LEVELING, []{queue.inject(F("G29N")); corner_probing_done = true;}, queue.inject(FPSTR(G28_STR))); ui.goto_previous_screen_no_defer();}
+      , []{
+        TERN(HAS_LEVELING, ui.goto_previous_screen_no_defer(), []{});
```
The anonymous empty lambda is not meaningful here. Instead it should be:
```cpp
TERN_(HAS_LEVELING, ui.goto_previous_screen_no_defer());
```

thinkyhead commented on Marlin/src/lcd/menu/menu_bed_corners.cpp:
```diff
@@ -311,21 +319,35 @@ static void _lcd_level_bed_corners_get_next_position() {
-  _lcd_draw_homing();
+void _lcd_level_bed_corners_homing() {
+  #if HAS_STOWABLE_PROBE && DISABLED(BLTOUCH)
+  if (!all_axes_homed() && probe.deploy) return;
```
- The value of `probe.deploy` is the address of the method, and it is always non-zero.
- The `_lcd_level_bed_corners_homing` screen is meant to stay visible until homing is done, as a stepping stone to the next screen handler, so it goes to a new screen handler as soon as homing has been completed. Since it is invoked as a screen handler, it must exit quickly. The `_lcd_test_corners` call sets a new screen handler, but it also runs a loop that takes over for the main loop. It's convoluted, but it works.

thinkyhead committed:
Don't use 'N' as a G-code word

thinkyhead committed:
probe.* cleanup

thinkyhead commented on Marlin/src/lcd/menu/menu_bed_corners.cpp:
```diff
@@ -346,20 +368,44 @@ static void _lcd_level_bed_corners_homing() {
+
+#endif
+
+void _lcd_level_bed_corners() {
```
Note: This method is a stepping stone to `_lcd_test_corners` (or the anonymous manual tramming handler), by way of `_lcd_level_bed_corners_homing`.

thinkyhead commented on Marlin/src/lcd/menu/menu_bed_corners.cpp:
```diff
@@ -346,20 +368,44 @@ static void _lcd_level_bed_corners_homing() {
+    _lcd_draw_homing();
+    if (all_axes_homed())
+      #if HAS_STOWABLE_PROBE && DISABLED(BLTOUCH)
+      deploy_probe();
```
Since this is within a screen handler, `deploy_probe` will be called repeatedly once axes are homed. Then there is nothing within this screen handler that will ever exit the screen handler, so it will keep doing this forever.

thinkyhead commented on Marlin/src/lcd/menu/menu_bed_corners.cpp:
```diff
@@ -346,20 +368,44 @@ static void _lcd_level_bed_corners_homing() {
+void deploy_probe() {
+  if (!corner_probing_done) probe.deploy(true);
+  ui.goto_screen([]{
+    _deploy_probe();
```
Since this is within a screen handler, `_deploy_probe` will be called repeatedly. Then there is nothing within this screen handler that exits the screen handler, so it will keep doing this until the code below invokes `_lcd_level_bed_corners_homing` to potentially set some new screen handler.

thinkyhead committed:
misc formatting

thinkyhead committed:
distinct *_leveling_was_active

thinkyhead committed:
escape from _lcd_level_bed_corners

thinkyhead committed:
fix up _lcd_draw_level_prompt

thinkyhead committed:
misc. followup

thinkyhead committed:
Distinguish DONE from BACK

thinkyhead committed:
deploy_probe() calls ui.goto…

thinkyhead commented:
I did a number of cleanups and tweaked a few things.

In general, the `BED_TRAMMING_Z_HOP` should be defined as a value large enough to clear the bed even if it's tilted by a mm or two, and `current_position.z + BED_TRAMMING_Z_HOP` doesn't necessarily give extra clearance because the bed might actually be lower at that point, so I suggest we use `BED_TRAMMING_Z_HOP` where we now use `current_position.z + BED_TRAMMING_Z_HOP`.

thinkyhead committed:
avoid low hop

Farva42 commented:
I tested all updates and everything still works. The only thing that might want to be changed is the STOW_BETWEEN_PROBE define as the probes requiring this are usually deployed once for the whole process and this is setup to allow that initial deploy and stow.

thinkyhead commented:
> I tested all updates and everything still works. The only thing that might want to be changed is the STOW_BETWEEN_PROBE define as the probes requiring this are usually deployed once for the whole process and this is setup to allow that initial deploy and stow.

Well, the `STOW_BETWEEN_PROBES` define is just shorthand for the conditions that were already there, namely `HAS_STOWABLE_PROBE && DISABLED(BLTOUCH)`. As I recall the BLTouch probe may or may not be in a mode where it automatically re-deploys the probe after it triggers (`BLTOUCH_HS_MODE` plus `M401 S1`), and when it is in this mode `bltouch.z_extra_clearance` will return 7.

So, if that makes any difference to the `_lcd_test_corners` function, now is a good time to make those tweaks.

Farva42 commented:
> > I tested all updates and everything still works. The only thing that might want to be changed is the STOW_BETWEEN_PROBE define as the probes requiring this are usually deployed once for the whole process and this is setup to allow that initial deploy and stow.
> 
> Well, the `STOW_BETWEEN_PROBES` define is just shorthand for the conditions that were already there, namely `HAS_STOWABLE_PROBE && DISABLED(BLTOUCH)`. As I recall the BLTouch probe may or may not be in a mode where it automatically re-deploys the probe after it triggers (`BLTOUCH_HS_MODE` plus `M401 S1`), and when it is in this mode `bltouch.z_extra_clearance` will return 7.
> 
> So, if that makes any difference to the `_lcd_test_corners` function, now is a good time to make those tweaks.

I had to add the conditions as the BLTOUCH is caught under the HAS_STOWABLE_PROBE via this define in Conditionals_lcd.h
```
#if ENABLED(BLTOUCH)
  #ifndef Z_PROBE_SERVO_NR
    #define Z_PROBE_SERVO_NR 0
  #endif
#endif

/**
 * Set a flag for a servo probe (or BLTouch)
 */
#ifdef Z_PROBE_SERVO_NR
  #define HAS_Z_SERVO_PROBE 1
```

The BLTOUCH has no probe.deploy or stow defined as it uses it's own calls so it will cause an error unless it's excluded from the deploy and stow calls. There's no functional issue with the naming just me being a little nitpicky to help with readability of the code.

Farva42 committed:
Change STOW_BETWEEN_PROBES

Farva42 committed:
Add current_position back in

Removal caused it to miss the centre point and try and probe upwards.

thinkyhead committed:
rename

thinkyhead committed:
cleanup

thinkyhead commented:
> The `BLTOUCH` has no `probe.deploy` or `stow` defined as it uses [its] own calls so it will cause an error unless it's excluded from the `deploy` and `stow` calls. There's no functional issue with the naming just me being a little nitpicky to help with readability of the code.

Well just to be clear, the `probe.deploy` / `stow` / `set_deployed` methods do the handling of BLTouch deploy and stow so that you don't need a special case for `BLTOUCH`; you just call the `probe` methods. For all kinds of probes, `set_deployed` handles any needed raise in a standard way before deploy/stow. Given that we have already tried to make all probes operate on the same level by calling through the `Probe` class, I would like to avoid making special allowances for the `BLTOUCH` within this one specific menu procedure.

In the testing of your bed tramming changes, it would be good to make sure that everything still works after both `M401 S0` and `M401 S1`, since the behavior of the BLTouch will change depending on the `bltouch.high_speed_mode` flag, and also test that everything still works when using a non-BLTouch probe. You should be able to fake a non-BLTouch probe while using a BLTouch by setting `Z_PROBE_SERVO_NR` to `0` and `Z_SERVO_ANGLES` to `{ 10, 90 }`.

Farva42 commented:
I might have been reading the log incorrectly but the test prior to adding the exclusion of the BLTOUCH it kicked out the following. 

```
[Test STM32F103VE_longer] BLTouch U20 config...
Marlin/src/lcd/menu/menu_bed_corners.cpp: In function 'void _lcd_level_bed_corners_homing()':
Marlin/src/lcd/menu/menu_bed_corners.cpp:331:28: error: 'probe' was not declared in this scope
  331 |   if (!all_axes_homed() && probe.deploy) return;
```

I'm also just realizing that with that test probe should still get defined so is it maybe an issue with a missing include?

I've been testing using a BLTOUCH as well as a mag_mounted_probe

Farva42 commented:
Ahh I found the issue. It was calling for a probe deploy when tram with probe was not enabled. I'm going to need to fix the needs deploy define to correctly fix that.

Farva42 committed:
Catch fringe case

This will stop issues if Bed_tramming is enabled with a probe but without tramming using probe.

Farva42 commented:
This should be good to go now. If wanted the ```DISABLED(BLTOUCH)```  can come out of the NEEDS_PROBE_DEPLOY if wanted but keeping it will mean that it would only deploy before probing instead of potentially being deployed while moving.

thinkyhead committed:
🎨 Misc. formatting

thinkyhead committed:
use BOTH

thinkyhead committed:
rename var

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25425

thinkyhead committed:
tweaks

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25425

thinkyhead committed:
expand _lcd_goto_next_corner

thinkyhead commented:
It made sense to merge `_lcd_bed_tramming_get_next_position` and `_lcd_goto_next_corner` which also saves some SRAM by removing the global `corner_point`. Hopefully there are no egregious typos. With the other neutral refinements already merged, this PR is reduced down to the logical changes related to `NEEDS_PROBE_DEPLOY`. Please confirm that it is still behaving as it was.

I have also been working on general probe refinements, but of course this procedure is different from regular probing, so there are no standard `Probe` class methods to handle its quirks. The main thing I'm still a little hung up on is when to use `line_to_z(current_position.z + BED_TRAMMING_Z_HOP)` versus just `BED_TRAMMING_Z_HOP`. It used to be that when `BED_TRAMMING_USE_PROBE` is enabled `_lcd_goto_next_corner` adds `current_position.z`, and otherwise it doesn't. So I've altered `_lcd_goto_next_corner` to always add `current_position.z` but maybe that is inappropriate. It would be worth comparing the behavior with and without `BED_TRAMMING_USE_PROBE` to make sure the raise amounts are appropriate in each case. I will test this tomorrow evening if you haven't already given feedback.

Farva42 commented:
I just got some time to test this. It is now causing looping with use probe enabled it will never exit and just continues to probe over and over. It is also not prompting to raise the bed if a corner is low it will just probe down to where it last hit and then go back up and go to the next corner. I'm going to start testing to see what caused it to start looping.

Farva42 committed:
Fix looping & pull current_position.z from hop

Farva42 commented:
I found the cause of the looping. The line_to_z move down to BED_TRAMMING_HEIGHT was causing the probe to be triggered and causing issues. 

In regards to the Tramming hop, having the current position added in was causing a massive hop for me as it would be adding it to the Z height after stowing the probe from the homing which for my setup is 20mm so my hop would be 24mm every time.  It could be changed to use Z_CLEARANCE_BETWEEN_PROBES as that should be a safe value as it would already be used between other probings. But I think having the current position can cause some excessive hops in some situations.

thinkyhead committed:
final touch

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25425
